### DISCUSSION TITLE
Breaking changes / updates on  WLX/Slides

### DISCUSSION THREAD

**Discussion by JerryI** [WLJSTeam/wljs-notebook#60] [Documentation]
It's been a long time we started to work on WLX ([Wolfram Language XML](https://jerryi.github.io/wljs-docs/wlx)) syntax extension. For now it entered into its final form.

Let me remind you, what you can do with it

```jsx
.wlx
<h1><RandomWord/></h1>
```

or in slides

```jsx
.wlx

Columns[Col1_, Col2_] := With[{},
    <div style="display: flex">
        <div style="width: 50%"><Col1/></div>
        <div style="width: 50%"><Col2/></div>
    </div>                       
]
```

```md
.slide

# Title

<Columns>
  <div>
    
## Subtitle 1

And some content
    
  </div>
  <div>

## Subtitle 2

And some content

  </div>
</Columns>
```

Now you have a component to make two columns

<img width="572" alt="Screenshot 2023-11-23 at 15 08 20-a41d8f5ca8e4850cfdb668e919cf3e94" src="https://github.com/JerryI/wolfram-js-frontend/assets/4111822/c8c60fc3-97a2-410e-a3bc-366f3cce4225">

👉 **See more on slides in the advanced section of docs ([SLIDES](https://jerryi.github.io/wljs-docs/docs/frontend/Advanced/Slides/intro))**

## Animations
In general all visuals can be done in the same way as in regular cells, since it uses the same approach

layout
```jsx
.slide

# Animation example

<Figure id={"routed-event-fragment-1"}/>

The figure will be changed, when the fragment below is revealed

<span style="color:red">Magic</span> <!-- .element: class="fragment" data-fragment-index="1" -->

<SlideEvent id={"routed-event"}/>
```

This is a typical approach I (@JerryI ) use for my reports on a work. Here we have an graphics component `Figure` with a parameter that specifies an `uid` for events, on which it is subscribed. And the event listener `SlideEvent`, that captures what happens on a slide (a slide got revealed or a fragment popped up) and sends them to an event object labelled as `"routed-event"`. 

Now the actual code for all functions we used

```mathematica
Figure[OptionsPatten[]] := With[{event = EventClone[OptionValue["id"]]},
   (* clean up, if the cell got destroyed *)
    EventHandler[EvaluationCell[], {"destroy" -> Function[Null, Delete[event]]}];
    
    LeakyModule[{points},
        (* initial state *)
        points = RandomReal[{-1,1}, {40,2}];
        
        EventHandler[event, Function[Null, 
            (* act when the event happend *)
            points = {Sin[#], Cos[#]} &/@ Range[40]
        ]];
        
        Graphics[{Red, Line[points // Offload]}]
    ]
]

Options[Figure] = {"id"->""}
```

This little script will plot randomly distributed points as lines for its initial state. When the event is fired, it changes the distribution of points to a circle. The result will look like

![ezgif com-video-to-gif-4c8934f8dec9251a05356f718152b388](https://github.com/JerryI/wolfram-js-frontend/assets/4111822/df0a1012-3a27-4921-a3e6-7e7be55a7005)

## What was broken?
We have changed the way, how parmeters or options are provided. See on the example below

```jsx
.wlx

<Heading Title={"Some title"}/>
```

where
__before__
```jsx
.wlx
Heading := With[{},
  <dummy>
    <h1><Title/></h1>
    Some repetative text you need
  </dummy>
] 
```

__after__

```jsx
.wlx
Heading[OptionsPattern[]]] := With[{Title = OptionValue["Title"]},
  <dummy>
    <h1><Title/></h1>
    Some repetative text you need
  </dummy>
] 

Options[Heading] = {"Title"->"Nope"}
```

Basically, before all parameters were passed as global variables, which is not obvious from the code structure. However, now it works as if it was usual `Options` of a symbol. Which is also great for WL ecosystem, since it bridges two languages nicer and provides clear view on what code does (comes handy if you many many of those pieces).

Cheers,
Kirill (@JerryI)

**Comment by JerryI** [comment]
UPD: All examples, packages and tutorials were updated accordingly
